File: platform/polyglot/figma_api/py/figma_api/logger.py
"""
Logger Module — Figma API SDK

Structured logging factory with sensitive data redaction.
Each package creates a scoped logger via `create_logger(package_name, __file__)`.

Usage:
    from .logger import create_logger
    log = create_logger("figma-api", __file__)
    log.info("fetching projects", team_id="123")
"""

import logging
import os
import sys
from pathlib import Path
from typing import Any, Optional

REDACT_KEYS = {"token", "secret", "password", "auth", "credential",
               "authorization", "apikey", "api_key", "accesstoken", "access_token"}

LEVELS = {
    "TRACE": 5,
    "DEBUG": logging.DEBUG,
    "INFO": logging.INFO,
    "WARN": logging.WARNING,
    "WARNING": logging.WARNING,
    "ERROR": logging.ERROR,
    "CRITICAL": logging.CRITICAL,
    "SILENT": 100,
}


def _redact_value(key: str, value: Any) -> Any:
    if isinstance(key, str) and key.lower() in REDACT_KEYS:
        if isinstance(value, str) and len(value) > 8:
            return value[:8] + "***"
        return "***"
    return value


def _format_kwargs(**kwargs: Any) -> str:
    if not kwargs:
        return ""
    parts = []
    for k, v in kwargs.items():
        v = _redact_value(k, v)
        parts.append(f"{k}={v!r}")
    return " " + " ".join(parts)


class SDKLogger:
    """SDK Logger with defensive programming patterns and credential redaction."""

    def __init__(
        self,
        package_name: str,
        filename: str,
        logger_instance: Optional[logging.Logger] = None,
    ):
        self.package_name = package_name
        self.filename = Path(filename).stem if filename else "unknown"
        self.prefix = f"[{package_name}:{self.filename}]"

        if logger_instance:
            self._logger = logger_instance
        else:
            self._logger = self._create_default_logger()

    def _create_default_logger(self) -> logging.Logger:
        logger = logging.getLogger(f"{self.package_name}.{self.filename}")
        if not logger.handlers:
            handler = logging.StreamHandler(sys.stderr)
            formatter = logging.Formatter(
                "%(asctime)s %(levelname)-5s %(message)s",
                datefmt="%Y-%m-%d %H:%M:%S",
            )
            handler.setFormatter(formatter)
            logger.addHandler(handler)

        level_name = os.environ.get("LOG_LEVEL", "INFO").upper()
        level = LEVELS.get(level_name, logging.INFO)
        logger.setLevel(level)
        return logger

    def _format(self, msg: str, **kwargs: Any) -> str:
        return f"{self.prefix} {msg}{_format_kwargs(**kwargs)}"

    def trace(self, msg: str, **kwargs: Any) -> None:
        self._logger.log(5, self._format(msg, **kwargs))

    def debug(self, msg: str, **kwargs: Any) -> None:
        self._logger.debug(self._format(msg, **kwargs))

    def info(self, msg: str, **kwargs: Any) -> None:
        self._logger.info(self._format(msg, **kwargs))

    def warn(self, msg: str, **kwargs: Any) -> None:
        self._logger.warning(self._format(msg, **kwargs))

    def warning(self, msg: str, **kwargs: Any) -> None:
        self.warn(msg, **kwargs)

    def error(self, msg: str, **kwargs: Any) -> None:
        self._logger.error(self._format(msg, **kwargs))

    def critical(self, msg: str, **kwargs: Any) -> None:
        self._logger.critical(self._format(msg, **kwargs))

    def set_level(self, level: str) -> None:
        level_value = LEVELS.get(level.upper(), logging.INFO)
        self._logger.setLevel(level_value)


def create_logger(
    package_name: str,
    filename: str,
    logger_instance: Optional[logging.Logger] = None,
) -> SDKLogger:
    """Create a logger instance for a module."""
    return SDKLogger(package_name, filename, logger_instance)

File: platform/polyglot/figma_api/mjs/src/logger.mjs
/**
 * Logger Module — Figma API SDK
 *
 * Structured logging factory with sensitive data redaction.
 * Each package creates a scoped logger via `create(packageName, filename)`.
 *
 * Usage:
 *   import { create } from './logger.mjs';
 *   const log = create('figma-api', import.meta.url);
 *   log.info('fetching projects', { teamId: '123' });
 */

import { fileURLToPath } from 'node:url';
import { basename } from 'node:path';

const LEVELS = { TRACE: 0, DEBUG: 1, INFO: 2, WARN: 3, ERROR: 4, SILENT: 100 };

const REDACT_KEYS = new Set([
  'token', 'secret', 'password', 'auth', 'credential',
  'authorization', 'apikey', 'api_key', 'accesstoken', 'access_token',
]);

function getLogLevel() {
  const name = (process.env.LOG_LEVEL || 'INFO').toUpperCase();
  return LEVELS[name] ?? LEVELS.INFO;
}

function formatTimestamp() {
  return new Date().toISOString().replace('T', ' ').split('.')[0];
}

function redactValue(key, value) {
  if (typeof key === 'string' && REDACT_KEYS.has(key.toLowerCase())) {
    if (typeof value === 'string' && value.length > 8) {
      return value.slice(0, 8) + '***';
    }
    return '***';
  }
  return value;
}

function redactContext(ctx) {
  if (!ctx || typeof ctx !== 'object') return ctx;
  const result = {};
  for (const [k, v] of Object.entries(ctx)) {
    result[k] = redactValue(k, v);
  }
  return result;
}

function formatContext(ctx) {
  if (!ctx || (typeof ctx === 'object' && Object.keys(ctx).length === 0)) return '';
  const redacted = redactContext(ctx);
  return ' ' + Object.entries(redacted)
    .map(([k, v]) => `${k}=${JSON.stringify(v)}`)
    .join(' ');
}

class SDKLogger {
  constructor(packageName, filename, customLogger = null) {
    this.packageName = packageName;
    if (filename) {
      try {
        const filePath = filename.startsWith('file://') ? fileURLToPath(filename) : filename;
        this.filename = basename(filePath).replace(/\.\w+$/, '');
      } catch {
        this.filename = 'unknown';
      }
    } else {
      this.filename = 'unknown';
    }
    this.prefix = `[${packageName}:${this.filename}]`;
    this.customLogger = customLogger;
  }

  _format(msg, ctx) {
    return `${formatTimestamp()} ${this.prefix} ${msg}${formatContext(ctx)}`;
  }

  _log(level, levelName, msg, ctx) {
    if (level < getLogLevel()) return;
    const formatted = this._format(msg, ctx);
    if (this.customLogger && typeof this.customLogger[levelName.toLowerCase()] === 'function') {
      this.customLogger[levelName.toLowerCase()](formatted);
      return;
    }
    switch (levelName) {
      case 'TRACE': console.debug(`TRACE ${formatted}`); break;
      case 'DEBUG': console.debug(`DEBUG ${formatted}`); break;
      case 'INFO': console.info(`INFO  ${formatted}`); break;
      case 'WARN': console.warn(`WARN  ${formatted}`); break;
      case 'ERROR': console.error(`ERROR ${formatted}`); break;
      default: console.log(`${levelName} ${formatted}`);
    }
  }

  trace(msg, ctx) { this._log(LEVELS.TRACE, 'TRACE', msg, ctx); }
  debug(msg, ctx) { this._log(LEVELS.DEBUG, 'DEBUG', msg, ctx); }
  info(msg, ctx) { this._log(LEVELS.INFO, 'INFO', msg, ctx); }
  warn(msg, ctx) { this._log(LEVELS.WARN, 'WARN', msg, ctx); }
  error(msg, ctx) { this._log(LEVELS.ERROR, 'ERROR', msg, ctx); }
}

export function create(packageName, filename, customLogger = null) {
  return new SDKLogger(packageName, filename, customLogger);
}

export { SDKLogger, LEVELS };
export default { create, SDKLogger, LEVELS };
